Objects/weakrefobject.c

```diff
@@ -1016,7 +1016,13 @@ PyObject_ClearWeakRefs(PyObject *object)
     PyObject *exc = PyErr_GetRaisedException();
     PyObject *tuple = PyTuple_New(num_weakrefs * 2);
     if (tuple == NULL) {
-        _PyErr_ChainExceptions1(exc);
+        _PyWeakref_ClearWeakRefsExceptCallbacks(object);
+        if (exc == NULL) {
+            PyErr_WriteUnraisable(NULL);
+        }
+        else {
+            _PyErr_ChainExceptions1(exc);
+        }
```

Comment:
Hmm... I think this should instead be:

```suggestion
        PyErr_WriteUnraisable(NULL);
        PyErr_SetRaisedException(exc);
```

In other words, write the unraisable exception and then restore the previous active exception, if any.

I don't think we should chain exceptions. The assertion in `_Py_Dealloc` checks that the exception did not change, but as written we would end up changing the exception if there was previously an active one.

Reply:
For an example of this pattern see _PyGen_Finalize:

https://github.com/python/cpython/blob/23d0371bb99b1df183c36883e256f82fdf6a4bea/Objects/genobject.c#L81-L89